cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -437,34 +425,39 @@ private void checkReconfigurability(PodRef podRef, Pod pod, RestartContext resta
                         && "Unschedulable".equals(ps.getReason())
                         && "False".equals(ps.getStatus()));
         if (podStuck
-                && !reasonToRestartPod.contains("Pod has old generation")   // "Pod has old generation" is used with StatefulSets
-                && !reasonToRestartPod.contains("Pod has old revision")) {  // "Pod has old revision" is used with PodSets
+                && !reasonsToRestartPod.contains(RestartReason.POD_HAS_OLD_GENERATION) // "Pod has old generation" is used with StatefulSets
+                && !reasonsToRestartPod.contains(RestartReason.POD_HAS_OLD_REVISION)) {  // "Pod has old revision" is used with PodSets
             // If the pod is unschedulable then deleting it, or trying to open an Admin client to it will make no difference
             // Treat this as fatal because if it's not possible to schedule one pod then it's likely that proceeding
             // and deleting a different pod in the meantime will likely result in another unschedulable pod.
             throw new FatalProblem("Pod is unschedulable");
         }
         // Unless the annotation is present, check the pod is at least ready.
-        boolean needsRestart = !reasonToRestartPod.isEmpty();
+        boolean needsRestart = reasonsToRestartPod.shouldRoll();
         KafkaBrokerConfigurationDiff diff = null;
         KafkaBrokerLoggingConfigurationDiff loggingDiff = null;
         boolean needsReconfig = false;
         // Always get the broker config. This request gets sent to that specific broker, so it's a proof that we can
         // connect to the broker and that it's capable of responding.
         if (!initAdminClient()) {
             LOGGER.infoCr(reconciliation, "Pod {} needs to be restarted, because it does not seem to responding to connection attempts", podRef);
-            restartContext.needsRestart = false;
-            restartContext.needsReconfig = false;
-            restartContext.forceRestart = true;
-            restartContext.diff = null;
-            restartContext.logDiff = null;
-            return;
+            kafkaRestartContext.setNeedsRestart(false);
+            kafkaRestartContext.setNeedsReconfig(false);
+
+            //TODO - figure out a way to handle this in a nicer manner
+            kafkaRestartContext.setForceRestart(true);
```

Comment:
Myabe don't expose these as setters but a richer API `kafkaRestartContext.markForceRestart(RestartReason.POD_UNRESPONSIVE)`

which internally sets `forceRestart=true` and clears the diff fields? I'm not sire if it should handle adding the restart reason itself or not...

Reply:
There's a lot more work needed in the roller, but it's out of scope for this PR :)